### PR TITLE
Remove fast faces

### PR DESCRIPTION
**Pull request by numberZero**
Reincarnation of #6903 and #13204.

Includes https://github.com/numberZero/minetest/tree/refactor-cuboid-drawing

@lhofhansl 

## To do

This PR is Ready for Review.

## How to test

> Load any world and test. Look at FPS and jitter.

Test with smooth and flat lighting.

### CODE REVIEW

File: src/client/content_mapblock.cpp [hunk 1]
```diff
@@ -404,9 +384,99 @@ void MapblockMeshGenerator::drawAutoLightedCuboid(aabb3f box, const f32 *txc,
...
+				layer.material_flags |= MATERIAL_FLAG_BACKFACE_CULLING;
+			else
+				layer.material_flags &= ~MATERIAL_FLAG_BACKFACE_CULLING;
+			layer.material_flags |= MATERIAL_FLAG_TILEABLE_HORIZONTAL;
+			layer.material_flags |= MATERIAL_FLAG_TILEABLE_VERTICAL;
+		}
+	}
+	if (!faces)
+		return;
+	u8 mask = faces ^ 0b0011'1111; // k-th bit is set if k-th face is to be *omitted*, as expected by cuboid drawing functions.
```
Comment (sfan5) [inline src/client/content_mapblock.cpp:457]: oh neat, binary literals are in C++14 apparently

File: src/client/content_mapblock.cpp [hunk 2]
```diff
@@ -404,9 +384,99 @@ void MapblockMeshGenerator::drawAutoLightedCuboid(aabb3f box, const f32 *txc,
...
+				video::S3DVertex &vertex = vertices[j];
+				vertex.Color = color;
+			}
+		});
+	}
+}
+
+void MapblockMeshGenerator::drawSolidNode()
+{
+	int faces = 0; // k-th bit will be set if k-th face is to be drawn.
```
Comment (sfan5) [inline src/client/content_mapblock.cpp:413]: `u8`

**Comment by sfan5** [comment]
For the uninformed people like me can we get a short explanation what fast faces does, why it was good and why it's no longer good now?

**Comment by lhofhansl** [comment]
Thanks @numberZero !!

@sfan5 FastFaces essentially do tile-combining. There's a scan along each principal axis and if the material, etc, are the same the faces are combined into one tiling. (but only along the axis, it will not combine across two axis' together)

This is less important now due to three reasons:
1. We now have mesh-chunking (compliments of @x2048)
2. Tiling was already disabled for waving water nodes
3. Tiling is generally disabled when dynamic shadows are enabled

On top of that it consumes considerable resources on the client - especially when mesh-chunking is enabled.
It's a hot mess of code. :)

File: src/client/content_mapblock.cpp (hunk 2 above)
Comment (numberZero) [reply src/client/content_mapblock.cpp]: Updated. That saved 8 bytes of object code!

**Comment by lhofhansl** [comment]
Works nicely. Did not perceive any slowdowns.
Smooth and flat lighting works fine (tried various worlds).

**Comment by lhofhansl** [comment]
Code looks reasonable. Good from my side. NB, it's a part of the code base that I understand well.

**Comment by lhofhansl** [comment]
If we are too worried about removing fast faces, would people agree to make this configurable?
Some duplicate code, but folks could try. After a release we could decide to keep or remove them.

I'm fine just dumping them now. I do have a nice Laptop, but it is by no means a fast gaming machine.

**Comment by ndren** [comment]
Tested on my old integrated graphics and it performs approximately the same (slightly worse, but not clear if that's because it started throttling after compiling). If it makes future improvements easier, go for it!

**Comment by lhofhansl** [comment]
I also tried with my Intel integrated card. With shadows and without, with mesh-chunking set to 1 and to 4, and I saw no measurable difference. My laptop is still pretty fast, but it's from 2019 and especially with the Intel integrated card not fancy in any way.

My vote is to rid ourselves of the FastFace mess.

**Comment by lhofhansl** [comment]
Same results on my old laptop from 2013, also with Intel integrated graphics.

Why do we even have FastFaces now? Am I missing something? Did we break something on the way? Or change something that obviates their need?

I see absolutely no measurable difference in performance.

**Comment by numberZero** [comment]
@lhofhansl FastFaces predate your 10 year old laptop.

**Comment by sfan5** [comment]
Windows build: [minetest-5.7.0-ceb6425-pr13216-win64.zip](https://github.com/minetest/minetest/files/10807300/minetest-5.7.0-ceb6425-pr13216-win64.zip)

Bug: `plantlike_rooted` nodes are missing their base

**Comment by numberZero** [comment]
Fixed. (dindn’t test).

**Comment by sfan5** [comment]
Tested:
* on my current system: no discernible performance differences
* laptop from 2011 (Intel HD Graphics 3000): not really anything either, maybe 2-3ms of drawtime?

the shading of node sides looks different:
screenshot_20230222_204425 before
screenshot_20230222_203529 after

**Comment by numberZero** [comment]
> the shading of node sides looks different:

If corners look the same, it may be due to using another diagonal for splitting a quad into triangles. Would be clearly visible in wireframe mode, and maybe also around light sources.

**Comment by numberZero** [comment]
Indeed there is a problem.
|FastFaces (before)|Cuboid (after)|
|--|--|
|FastFaces|Cuboid|

FastFaces had special handing to make lighting look symmetrical. It should be easy to replicate that to all cuboids (that would include nodeboxes).

Technically it is possible to make split direction not matter at all by shading with bilinear interpolation instead of barycentric one. But without GL3+ that’d be a BIG headache.

**Comment by lhofhansl** [comment]
I can hardly see the difference.
Is it just different tessellation? If so, one way is not more correct than another.
@numberZero you are saying there is some special logic we're using that used to reduce the effect of different tessellation?

**Comment by numberZero** [comment]
> you are saying there is some special logic we're using that used to reduce the effect of different tessellation?

Yes I am. Here is it:
https://github.com/minetest/minetest/blob/6e1c70e02b00df94a584bb4247cfa4f4c616c8d5/src/client/mapblock_mesh.cpp#L587-L588
https://github.com/minetest/minetest/blob/6e1c70e02b00df94a584bb4247cfa4f4c616c8d5/src/client/mapblock_mesh.cpp#L1234-L1237

**Comment by lhofhansl** [comment]
Looks like this is decided per tile... So would we just add the same logic to drawCuboid? I.e. check lights in the same way, then use either quad_indeces or the same alternative order used in FFs?

**Comment by numberZero** [comment]
> So would we just add the same logic to drawCuboid?

That’s what I meant:

> It should be easy to replicate that to all cuboids (that would include nodeboxes).

**Comment by lhofhansl** [comment]
That's code I'm afraid to touch. Also seems like I don't have the eye to catch these "minimal" rendering differences.
Are you planning to add a fix @numberZero . I think we're close to finally get rid of Fast Faces.

**Comment by numberZero** [comment]
Done.

File: src/client/content_mapblock.cpp [hunk 3]
```diff
@@ -56,7 +56,9 @@ static const v3s16 light_dirs[8] = {
 };
 
 // Standard index set to make a quad on 4 vertices
-static constexpr u16 quad_indices[] = {0, 1, 2, 2, 3, 0};
+static constexpr u16 quad_indices_02[] = {0, 1, 2, 2, 3, 0};
+static constexpr u16 quad_indices_13[] = {0, 1, 3, 3, 1, 2};
+static auto &quad_indices = quad_indices_02;
```
[the remark this answers is not in the document]
Comment (numberZero) [reply src/client/content_mapblock.cpp:61]: It’s const already. But if implicitness bothers you...

**Comment by lhofhansl** [comment]
I maintain my "One Approval"

**Comment by Desour** [comment]
For liquid sources, it would maybe be better to always use the same tesselation. If you enable waving liquids, set wavelength very short (i.e. 2), and then place a torch next to a lake, you can see differences.
(Not necessary for this PR though, as this already happens in master.)

Flowing water is not affected, as you still use the old `quad_indices`.
(This is not optimal though, as we currently have this shape asymmetry (could be fixed by always choosing the outer edge):
screenshot_20230224_201012
Anyway, this is somewhat offtopic.)

**Comment by x2048** [comment]
I think I've spotted a regression in how the smooth light is calculated / applied to the vertices:

Here's the shading on master:
image
image

Here's got it looks in the PR:
image
image

Here's how the wall looks from the other side:
image

**Comment by numberZero** [comment]
Good catch. It uses `getSmoothLightTransparent` to calculate light levels at node corners but for a solid node, there is no such thing as “light level at a corner,” it is face-dependent.

**Comment by numberZero** [comment]
Fixed. Now should work like in master. That does mean that a solid node and a full-node nodebox may look differently in the same position, though.

**Comment by lhofhansl** [comment]
I'd really like to get this in - before it bitrots :)

What can we do here? More testing? Make it configurable?

**Comment by numberZero** [comment]
> What can we do here? More testing? Make it configurable?

Review, test, merge)

File: src/client/content_mapblock.cpp [hunk 4]
```diff
@@ -366,6 +358,11 @@ void MapblockMeshGenerator::generateCuboidTextureCoords(const aabb3f &box, f32 *
 		coords[i] = txc[i];
 }
 
+int lightDiff(LightPair a, LightPair b)
```
Comment (sfan5) [inline src/client/content_mapblock.cpp:361]:
```suggestion
static inline int lightDiff(LightPair a, LightPair b)
```

**Comment by sfan5** [comment]
FWIW I consider this ready. One thing I just still wanted to do (which I haven't had time for) is compare the performance on Android

**Comment by x2048** [comment]
Core i5 3317U with integrated GPU, no performance difference.

**Comment by lhofhansl** [comment]
Any luck with Android @sfan5 . I'm actually quite curious about that.

**Comment by sfan5** [comment]
Inconclusive, results don't indicate either being faster (with gles2).
[mt-master.apk.zip](https://github.com/minetest/minetest/files/10902026/mt-master.apk.zip)
[mt-with-pr.apk.zip](https://github.com/minetest/minetest/files/10902032/mt-with-pr.apk.zip)

thinking out loud: it would be good if there was a game that put you in a controlled environment, viewpoint and scene to test rendering of a specific scene

**Comment by lhofhansl** [comment]
To see why this one is necessary try this test:

* set `client_mesh_chunk` to a large value (anything between 8 and 16)
* place or dig some blocks
* Notice the larger delay between your action and the time the change is actually visually reflected on the client
* This extra delay is due to time spent in mesh-generation

You'll find that this PR cuts the time in this feedback loop by about 50% (there is just much less work to do).

On my machine `client_mesh_chunk` = 10 is OK with this PR but useless without.
(16 is unacceptably slow even with this PR, but that is a different discussion.)

I'm sure there are other way to speed up mesh-generation. @x2048

**Review by sfan5 — APPROVED** [review]

**Comment by rubenwardy** [comment]
It feels to me that we're at a local optima and that perhaps a different geometry optimisation algorithm would improve performance more significantly. But does make sense to remove fast faces as-is

**Comment by rubenwardy** [comment]
Yeah, it's too risky to go in during feature freeze

File: src/client/content_mapblock.cpp (hunk 2 above)
Comment (numberZero) [reply src/client/content_mapblock.cpp]: Note: at -O3 my gcc optimizes it to `int` +)

**Comment by lhofhansl** [comment]
Merged! Yeah!

**Pull request merged**